### PR TITLE
Make use of CMakeDebHelper to improve CPack output

### PR DESCRIPTION
**Pull request by daschuer**
This is the next step towards getting rid of Scons. 

Missing: 
Retrieving the git version number 
Adding man page and other things to the deb file. 

This is PR is considered as ready .

### CODE REVIEW

**Review by Holzhaus — CHANGES_REQUESTED** [review]
Thanks for working on this, but I don't get what you are trying to archieve? Building DEBs already works. What is all this custom code supposed to do? Can't this be archieved with built-in CPack variables?

And why do we need the additional dependecy on debhelper? Right now I can build DEB packages on Arch without having debhelper installed.

> Adding man page and other things to the deb file.

Why add a man page just to the deb file? IMHO that should be added to the install commands that affect all unix systems. This should be autogenerated from some file, we shouldn't maintain documentation in that many places.

**Comment by daschuer** [comment]
> Building DEBs already works. 

I can confirm this on Ubuntu as well. However the Deb file only contains the bare minimum Infos to be usable.

> What is all this custom code supposed to do? Can't this be archieved with built-in CPack variables?

It currently sets up the menu. We can do it in other ways, but using the DebHelpers avoids to put the knowledge about the requirements of figernt distros into our source. We can assume that they are patched to work flawlessly today and tomorrow.
The two additional files in the cmake folders are nice helpers that I have found at GitHub.
It releases us from wrapping our heads around how to deal with the right folder structures. They just work.

Regarding the man file: I do not mind if and how we integrated it. I don't expect anyone using it. It was just up to now we added during the Deb file creation.

File: CMakeLists.txt [hunk 1]
```diff
@@ -993,12 +995,24 @@ set(CPACK_RESOURCE_FILE_README "${CMAKE_CURRENT_SOURCE_DIR}/README")
...
+set(CPACK_DEBIAN_PACKAGE_MAINTAINER "${CPACK_PACKAGE_CONTACT}")
 set(CPACK_DEBIAN_PACKAGE_SECTION "sound")
 set(CPACK_DEBIAN_PACKAGE_PRIORITY "optional")
 set(CPACK_DEBIAN_PACKAGE_SUGGESTS "pdf-viewer")
 set(CPACK_DEBIAN_PACKAGE_REPLACES "mixxx-data")
-set(CPACK_DEBIAN_PACKAGE_DEPENDS "libqt5opengl5, libqt5svg5, libqt5xml5, libqt5sql5, libqt5sql5-sqlite")
+set(CPACK_DEBIAN_PACKAGE_DEPENDS "libqt5sql5-sqlite, pulseaudio-utils")
 set(CPACK_DEBIAN_PACKAGE_SHLIBDEPS ON)
+set(CPACK_DEBIAN_PACKAGE_VERSION "2.4.0~alpha~pre-ppa1~master~git7371~bionic")
+set(CPACK_DEBIAN_PACKAGE_HOMEPAGE "${CPACK_PACKAGE_HOMEPAGE_URL}")
```
[the remark this answers is not in the document]
Comment (daschuer) [reply CMakeLists.txt:1006]: It does not work here unfortunately.
So I set it explicit.

File: CMakeLists.txt [hunk 2]
```diff
@@ -993,12 +995,24 @@ set(CPACK_RESOURCE_FILE_README "${CMAKE_CURRENT_SOURCE_DIR}/README")
...
+set(CPACK_DEBIAN_PACKAGE_DESCRIPTION
+"Digital Disc Jockey Interface
+ Mixxx is a digital DJ system, where Wave, Ogg, FLAC and MP3 files can be
+ mixed on a computer for use in live performances. Filters, crossfader, and
+ speed control are provided. Mixxx can sync the 2 streams automatically, using
+ an algorithm to detect the beat.
+ .
+ Mixxx works with Jack or OSS, can be controlled from the GUI or from external
+ controllers, including MIDI devices and joysticks, and supports skins.")
```
[the remark this answers is not in the document]
Comment (daschuer) [reply CMakeLists.txt:1015]: This also does not work by default.
I have tried to read the description file but that produced a broken deb.
Than I noticed that the debian maintainers rip our long description anyway. So I have decided to use the description of the debian maintainers directly. 
Hopefully this way it is back under our control.

File: CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (uklotzde) [reply CMakeLists.txt:1006]: Hard-coding distribution-specific build versions into the generic build file doesn't seem to be right.

[the remark this answers is not in the document]
Comment (daschuer) [reply CMakeLists.txt:1006]: Yes, I know. This is s place holder to generate the git version from cmake.
This is quite a bit work I want to avoid now.

File: CMakeLists.txt [hunk 3]
```diff
@@ -993,12 +995,24 @@ set(CPACK_RESOURCE_FILE_README "${CMAKE_CURRENT_SOURCE_DIR}/README")
...
 set(CPACK_DEBIAN_PACKAGE_PRIORITY "optional")
 set(CPACK_DEBIAN_PACKAGE_SUGGESTS "pdf-viewer")
 set(CPACK_DEBIAN_PACKAGE_REPLACES "mixxx-data")
-set(CPACK_DEBIAN_PACKAGE_DEPENDS "libqt5opengl5, libqt5svg5, libqt5xml5, libqt5sql5, libqt5sql5-sqlite")
+set(CPACK_DEBIAN_PACKAGE_DEPENDS "libqt5sql5-sqlite, pulseaudio-utils")
 set(CPACK_DEBIAN_PACKAGE_SHLIBDEPS ON)
+set(CPACK_DEBIAN_PACKAGE_VERSION "2.4.0~alpha~pre-ppa1~master~git7371~bionic")
+set(CPACK_DEBIAN_PACKAGE_HOMEPAGE "${CPACK_PACKAGE_HOMEPAGE_URL}")
+set(CPACK_DEBIAN_PACKAGE_DESCRIPTION
+"Digital Disc Jockey Interface
```
Comment (Be-ing) [inline CMakeLists.txt:1008]:
```suggestion
"Digital Disc Jockey Application
```
?

File: CMakeLists.txt [hunk 4]
```diff
@@ -993,12 +995,24 @@ set(CPACK_RESOURCE_FILE_README "${CMAKE_CURRENT_SOURCE_DIR}/README")
...
+ Mixxx is a digital DJ system, where Wave, Ogg, FLAC and MP3 files can be
+ mixed on a computer for use in live performances. Filters, crossfader, and
+ speed control are provided. Mixxx can sync the 2 streams automatically, using
```
Comment (Be-ing) [inline CMakeLists.txt:1011]:
```suggestion
 speed control are provided. Mixxx can sync the 4 decks automatically, using
```

File: CMakeLists.txt [hunk 5]
```diff
@@ -993,12 +995,24 @@ set(CPACK_RESOURCE_FILE_README "${CMAKE_CURRENT_SOURCE_DIR}/README")
...
+ an algorithm to detect the beat.
+ .
```
Comment (Be-ing) [inline CMakeLists.txt:1013]: Does this `.` need to be here?

File: CMakeLists.txt [hunk 6]
```diff
@@ -993,12 +995,24 @@ set(CPACK_RESOURCE_FILE_README "${CMAKE_CURRENT_SOURCE_DIR}/README")
...
+ Mixxx works with Jack or OSS, can be controlled from the GUI or from external
```
Comment (Be-ing) [inline CMakeLists.txt:1014]:
```suggestion
 Mixxx works with ALSA or JACK, can be controlled from the GUI or from external
```

File: CMakeLists.txt (hunk 5 above)
Comment (daschuer) [reply CMakeLists.txt:1013]: Yes, that is a debian hack for an empty line.

File: CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (Holzhaus) [reply CMakeLists.txt:1015]: Can you update the description file and use `file(READ ...)` to read the file contents into a variable? That should work around the description file issue and keeps the description in a single place.

[the remark this answers is not in the document]
Comment (daschuer) [reply CMakeLists.txt:1015]: You mean the short description for all?

File: CMakeLists.txt [hunk 7]
```diff
@@ -993,12 +995,24 @@ set(CPACK_RESOURCE_FILE_README "${CMAKE_CURRENT_SOURCE_DIR}/README")
...
 
+set(CPACK_DEBIAN_PACKAGE_MAINTAINER "${CPACK_PACKAGE_CONTACT}")
 set(CPACK_DEBIAN_PACKAGE_SECTION "sound")
 set(CPACK_DEBIAN_PACKAGE_PRIORITY "optional")
 set(CPACK_DEBIAN_PACKAGE_SUGGESTS "pdf-viewer")
 set(CPACK_DEBIAN_PACKAGE_REPLACES "mixxx-data")
-set(CPACK_DEBIAN_PACKAGE_DEPENDS "libqt5opengl5, libqt5svg5, libqt5xml5, libqt5sql5, libqt5sql5-sqlite")
+set(CPACK_DEBIAN_PACKAGE_DEPENDS "libqt5sql5-sqlite, pulseaudio-utils")
 set(CPACK_DEBIAN_PACKAGE_SHLIBDEPS ON)
+set(CPACK_DEBIAN_PACKAGE_VERSION "2.4.0~alpha~pre-ppa1~master~git7371~bionic")
```
[the remark this answers is not in the document]
Comment (daschuer) [reply CMakeLists.txt:1005]: We need stick with the curious version string for our deb file to fit into the > check. 
This doe not make sense for our other builds target. 
Fixing this should be done in a separate PR. The big underlying issue is that the git version is calculated at configure time, but not at pack time. 
We need a solution that uses the version from the binary that is packed and not an outdated version from the configure time. 
Do you know how to achieve this?

**Comment by daschuer** [comment]
All green, finally.

File: CMakeLists.txt [hunk 8]
```diff
@@ -980,6 +980,8 @@ if(UNIX AND NOT APPLE)
 endif()
 
 # Packaging
+
+set(CPACK_PACKAGE_NAME "mixxx")
```
Comment (Holzhaus) [inline CMakeLists.txt:984]: This defaults to the project name, so we could remove this as it's unneeded.

File: CMakeLists.txt [hunk 9]
```diff
@@ -993,12 +995,24 @@ set(CPACK_RESOURCE_FILE_README "${CMAKE_CURRENT_SOURCE_DIR}/README")
 set(CPACK_STRIP_FILES ON)
 set(CPACK_CREATE_DESKTOP_LINKS "mixxx")
 
+set(CPACK_DEBIAN_PACKAGE_MAINTAINER "${CPACK_PACKAGE_CONTACT}")
```
[the remark this answers is not in the document]
Comment (Holzhaus) [reply CMakeLists.txt:998]: ping

File: CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (Holzhaus) [reply CMakeLists.txt:1015]: > You mean the short description for all?

I don't understand your question, can you rephrase?

File: CMakeLists.txt [hunk 10]
```diff
@@ -1007,6 +1021,13 @@ set(CPACK_WIX_PROPERTY_ARPHELPLINK "${CPACK_PACKAGE_HOMEPAGE_URL}")
 set(CPACK_WIX_UI_BANNER "${CMAKE_CURRENT_SOURCE_DIR}/build/wix/images/banner.bmp")
 set(CPACK_WIX_UI_DIALOG "${CMAKE_CURRENT_SOURCE_DIR}/build/wix/images/dialog.bmp")
 
+if(UNIX AND NOT APPLE)
+  add_subdirectory(build/debian)
+  add_subdirectory(build/rpm)
+endif()
```
Comment (Holzhaus) [inline CMakeLists.txt:1027]: The code from these `CMakeLists.txt` files should go into a `CPackConfigDEB.cmake` and `CPackConfigRPM.cmake` inside the `cmake` directory instead.

Then, you can include them from the `CPackConfig.cmake` file (not `CMakeLists.txt`) like this:

    if(CPACK_GENERATOR MATCHES "DEB")
        include(CPackConfigDEB)
    endif()

**Review by Holzhaus — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (daschuer) [reply CMakeLists.txt:1015]: Do you mean we should discard our long description and use the shorted description from the Debian package for all our packages?

File: CMakeLists.txt (hunk 10 above)
Comment (daschuer) [reply CMakeLists.txt:1027]: I am not sure if this works, because CPackConfig.cmake is used at packaging time. Did you try it? 

But anyway, the current structure is done in a way that the project specific settings are in CMakeList.txt files in context at the debian and rpm folder while the common files that can be used in any project live in the cmake folder. I would like to keep this. 

What is the bad thing with add_subdirectory here?

File: CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (Holzhaus) [reply CMakeLists.txt:1013]: Yes.

File: CMakeLists.txt (hunk 7 above)
[the remark this answers is not in the document]
Comment (Holzhaus) [reply CMakeLists.txt:1192]: `CPackConfig` runs at configure time. But I don't think we should overwrite this variable at all. What I want to build a package for debian?

Instead, we should invoke it like this: cpack -G -DCPACK_PACKAGE_VERSION=foo

File: CMakeLists.txt [hunk 11]
```diff
@@ -2445,38 +2451,11 @@ set(CPACK_WIX_PROPERTY_ARPHELPLINK "${CPACK_PACKAGE_HOMEPAGE_URL}")
 set(CPACK_WIX_UI_BANNER "${CMAKE_CURRENT_SOURCE_DIR}/build/wix/images/banner.bmp")
 set(CPACK_WIX_UI_DIALOG "${CMAKE_CURRENT_SOURCE_DIR}/build/wix/images/dialog.bmp")
 
-include(CPack)
-
-if(APPLE AND MACOS_BUNDLE)
```
Comment (Be-ing) [inline CMakeLists.txt:2450]: Why did you remove macOS bundle support?

Comment (daschuer) [reply CMakeLists.txt:2450]: Ups, that was not intended, a rebasing issue. I will rebase again.

File: build/debian/mixxx.menu.in [hunk 12]
```diff
@@ -0,0 +1,7 @@
+?package(mixxx): \
+  needs="X11" \
+  section="Apps/Sound" \
+  title="Mixxx" \
+  command="/usr/bin/pasuspender -- /usr/bin/mixxx -platform xcb" \
```
Comment (Be-ing) [inline build/debian/mixxx.menu.in:5]:
```suggestion
  command="/usr/bin/pasuspender -- /usr/bin/mixxx -platform xcb || /usr/bin/mixxx -platform xcb" \
```
This should not require PulseAudio to be installed.

File: CMakeLists.txt [hunk 13]
```diff
@@ -2419,7 +2419,8 @@ endif()
 # Packaging
 set(CPACK_PACKAGE_VENDOR "Mixxx Project")
 set(CPACK_PACKAGE_CONTACT "RJ Skerry-Ryan <rryan@mixxx.org>")
-set(CPACK_PACKAGE_DESCRIPTION_FILE "${CMAKE_CURRENT_SOURCE_DIR}/cmake/cpack_package_description.txt")
+set(CPACK_PACKAGE_DESCRIPTION_SUMMARY "Mixxx, powerfull features for all DJs")
```
Comment (Be-ing) [inline CMakeLists.txt:2422]:
```suggestion
set(CPACK_PACKAGE_DESCRIPTION_SUMMARY "Mixxx DJ software")
```

Comment (Be-ing) [reply CMakeLists.txt:2422]: No need for embellishing language here, just a plain description is more helpful IMO

File: build/debian/mixxx.menu.in (hunk 12 above)
Comment (Holzhaus) [reply build/debian/mixxx.menu.in:5]: I don't think is actually runs a full shell. And this automatically restarts mixxxx if it crashes or exists with non-zero status, which is confusing.

If pasuspender is installed by default on Ubuntu, I think it's okay to use it. If you really want the fallback, we should use a separate shell script that tests if that command is in `$PATH`.

**Comment by Be-ing** [comment]
I am still unclear what is the need for this. We don't really need to make binary DEB packages. We only need to making a source DEB to upload to Launchpad.

**Comment by daschuer** [comment]
Launchpad used the makefile debian/rules. This will call whatever necessary to create the Deb file.
I think it is cpack in our case. We will see how it turns out finally.

**Comment by Be-ing** [comment]
Is this still relevant with #3426?

**Comment by daschuer** [comment]
Yes if we want continue to use 
cpack -G DEB
Do we?

**Comment by Be-ing** [comment]
I don't think we should deploy DEB packages to downloads.mixxx.org and instead only direct users to the PPA. However it is nice to have DEB packages as build artifacts for PRs.

**Comment by Be-ing** [comment]
Please remove the `.travis.yml` file. I suppose that was accidentally reintroduced in a merge conflict resolution. If there are changes to CI that should be done for this, do them with GitHub Actions.

**Comment by daschuer** [comment]
superseded by https://github.com/mixxxdj/mixxx/pull/3571
